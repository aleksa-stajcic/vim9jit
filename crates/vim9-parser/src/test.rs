#![cfg(test)]

use crate::*;

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            setup_trace();

            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_parsing(contents));
            });
        }
    };
}

snapshot!(test_var, "../testdata/snapshots/simple_var.vim");
snapshot!(test_ret, "../testdata/snapshots/simple_ret.vim");
snapshot!(test_shared, "../testdata/snapshots/shared.vim");
snapshot!(test_comment, "../testdata/snapshots/comment.vim");
snapshot!(test_header, "../testdata/snapshots/header.vim");
snapshot!(test_expr, "../testdata/snapshots/expr.vim");
snapshot!(test_echo, "../testdata/snapshots/echo.vim");
snapshot!(test_scopes, "../testdata/snapshots/scopes.vim");
snapshot!(test_autocmd, "../testdata/snapshots/autocmd.vim");
snapshot!(test_array, "../testdata/snapshots/array.vim");
snapshot!(test_dict, "../testdata/snapshots/dict.vim");
snapshot!(test_if, "../testdata/snapshots/if.vim");
snapshot!(test_call, "../testdata/snapshots/call.vim");
snapshot!(test_concat, "../testdata/snapshots/concat.vim");
snapshot!(test_unpack, "../testdata/snapshots/unpack.vim");
snapshot!(test_assign, "../testdata/snapshots/assign.vim");
snapshot!(test_vimvar, "../testdata/snapshots/vimvar.vim");
snapshot!(test_busted, "../testdata/snapshots/busted.vim");
snapshot!(test_heredoc, "../testdata/snapshots/heredoc.vim");
snapshot!(test_typed_params, "../testdata/snapshots/typed_params.vim");
snapshot!(test_index, "../testdata/snapshots/index.vim");
snapshot!(test_adv_index, "../testdata/snapshots/adv_index.vim");
snapshot!(test_multiline, "../testdata/snapshots/multiline.vim");
snapshot!(test_cfilter, "../testdata/snapshots/cfilter.vim");
snapshot!(test_lambda, "../testdata/snapshots/lambda.vim");
snapshot!(test_comparisons, "../testdata/snapshots/comparisons.vim");
snapshot!(test_methods, "../testdata/snapshots/methods.vim");
snapshot!(test_eval, "../testdata/snapshots/eval.vim");
snapshot!(test_export, "../testdata/snapshots/export.vim");
snapshot!(test_import, "../testdata/snapshots/import.vim");
snapshot!(
    test_plugin_fileselect,
    "../testdata/snapshots/plugin_fileselect.vim"
);

#[test]
fn test_peek_n() {
    let input = "vim9script\nvar x = true\n";
    let lexer = new_lexer(input);
    let mut parser = Parser::new(lexer);
    assert_eq!(parser.peek_token.kind, TokenKind::EndOfLine);
    assert_eq!(parser.peek_n(0).kind, TokenKind::Identifier);
    assert_eq!(parser.peek_n(1).kind, TokenKind::EndOfLine);
    assert_eq!(parser.peek_n(2).kind, TokenKind::Identifier);
    assert_eq!(parser.peek_n(3).kind, TokenKind::Identifier);
    assert_eq!(parser.peek_n(4).kind, TokenKind::Equal);
    assert_eq!(parser.peek_n(1).kind, TokenKind::EndOfLine);
}

// TODO: Slowly but surely, we can work towards this
// snapshot!(test_matchparen, "../../shared/snapshots/matchparen.vim");
