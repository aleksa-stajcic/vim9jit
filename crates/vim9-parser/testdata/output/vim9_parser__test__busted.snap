---
source: crates/vim9-parser/src/test.rs
assertion_line: 38
expression: snapshot_parsing(contents)
---
[
    Vim9Script(
        Vim9ScriptCommand {
            noclear: false,
            eol: Token(EndOfLine, "\n", (0,10)->(0,10)),
        },
    ),
    Def(
        DefCommand {
            def: Token(Identifier, "def", (2,0)->(2,3)),
            name: Raw(Test_assignment_one),
            args: Signature {
                open: Token(LeftParen, "(", (2,23)->(2,24)),
                params: [],
                close: Token(RightParen, ")", (2,24)->(2,25)),
            },
            ret: None,
            def_eol: Token(EndOfLine, "\n", (2,25)->(2,25)),
            body: Body {
                commands: [
                    Var(
                        VarCommand {
                            var: Token(Identifier, "var", (3,2)->(3,5)),
                            ty: Some(
                                Type {
                                    colon: Token(SpacedColon, ": ", (3,11)->(3,13)),
                                    inner: Bool,
                                },
                            ),
                            name: Raw(bool1),
                            equal: Token(Equal, "=", (3,18)->(3,19)),
                            expr: Boolean(
                                VimBoolean {
                                    token: Token(True, "true", (3,20)->(3,24)),
                                    value: true,
                                },
                            ),
                            eol: Token(EndOfLine, "\n", (3,24)->(3,24)),
                        },
                    ),
                    Call(
                        CallCommand {
                            call: None,
                            expr: Identifier(
                                Raw(assert_equal),
                            ),
                            open: Token(LeftParen, "(", (4,14)->(4,15)),
                            args: [
                                Identifier(
                                    Scope(ScopedIdentifier { scope: VimVar, colon: Token(Colon, ":", (4,16)->(4,17)), accessor: Raw(true) }),
                                ),
                                Identifier(
                                    Raw(bool1),
                                ),
                            ],
                            close: Token(RightParen, ")", (4,28)->(4,29)),
                            eol: Token(EndOfLine, "\n", (4,29)->(4,29)),
                        },
                    ),
                ],
            },
            enddef: Token(Identifier, "enddef", (5,0)->(5,6)),
            end_eol: Token(EndOfLine, "\n", (5,6)->(5,6)),
        },
    ),
]
